curvefs/src/metaserver/dentry_storage.cpp

```diff
@@ -49,86 +50,148 @@ bool operator<(const Dentry& lhs, const Dentry& rhs) {
            LESS4(fsid, parentinodeid, name, txid);
 }
 
-DentryStorage::DentryStorage(std::shared_ptr<KVStorage> kvStorage,
-                             const std::string& tablename)
-    : kvStorage_(kvStorage),
-      tablename_(tablename) {}
-
-inline std::string DentryStorage::DentryKey(const Dentry& dentry,
-                                            bool ignoreTxId) {
-    std::ostringstream oss;
-    std::string txId = ignoreTxId ? "" : std::to_string(dentry.txid());
-    oss << dentry.fsid() << ":" << dentry.parentinodeid() << ":"
-        << Hash(dentry.name()) << ":" << txId;
-    return oss.str();
+static bool BelongSomeOne(const Dentry& lhs, const Dentry& rhs) {
+    return EQUAL(fsid) && EQUAL(parentinodeid) && EQUAL(name) &&
+           EQUAL(inodeid);
 }
 
-inline std::string DentryStorage::SameParentKey(const Dentry& dentry) {
-    std::ostringstream oss;
-    oss << dentry.fsid() << ":" << dentry.parentinodeid() << ":";
-    return oss.str();
+static bool IsSameDentry(const Dentry& lhs, const Dentry& rhs) {
+    return EQUAL(fsid) && EQUAL(parentinodeid) && EQUAL(name) &&
+           EQUAL(txid);
 }
 
-bool DentryStorage::BelongSameOne(const Dentry& lhs, const Dentry& rhs) {
-    return EQUAL(fsid) && EQUAL(parentinodeid) &&
-           EQUAL(name) && lhs.txid() <= rhs.txid();
+static bool HasDeleteMarkFlag(const Dentry& dentry) {
+    return (dentry.flag() & DentryFlag::DELETE_MARK_FLAG) != 0;
 }
 
-bool DentryStorage::IsSameDentry(const Dentry& lhs, const Dentry& rhs) {
-    return EQUAL(fsid) && EQUAL(parentinodeid) && EQUAL(name) &&
-           EQUAL(inodeid);
+static void VecFilter(BTree* btree, const DentryVec& vec, uint64_t maxTxId) {
+    for (const Dentry& dentry : vec.dentrys()) {
+        if (dentry.txid() <= maxTxId) {
+            btree->insert(dentry);
+        }
+    }
 }
 
-inline bool DentryStorage::HasDeleteMarkFlag(const Dentry& dentry) {
-    return (dentry.flag() & DentryFlag::DELETE_MARK_FLAG) != 0;
+static bool VecInsert(DentryVec* vec, const Dentry& dentry) {
+    for (const Dentry& item : vec->dentrys()) {
+        if (item == dentry) {
+            return false;
+        }
+    }
+    vec->add_dentrys()->CopyFrom(dentry);
+    return true;
+}
+
+static void VecDelete(DentryVec* vec, const Dentry& dentry) {
+    for (size_t i = 0; i < vec->dentrys_size(); i++) {
+        if (IsSameDentry(vec->dentrys(i), dentry)) {
+            vec->mutable_dentrys()->DeleteSubrange(i, 1);
+            break;
+        }
+    }
+}
+
+static void VecMerge(DentryVec* dst, const DentryVec& src) {
+    for (const auto& dentry : src.dentrys()) {
+        dst->add_dentrys()->CopyFrom(dentry);
```

Comment:
Do you need to remove duplicate dentry here?

Reply:
> 
no~ we need not~